### ISSUE TITLE
[Spike] Pre-requisite for fault-tolerant client-server 

### ISSUE THREAD

**Issue by mshabarov** [vaadin/flow#20418]
### Describe your motivation

This is a pre-requisite investigation task before we can start https://github.com/vaadin/flow/issues/20348.
The description there is not enough to start the development and it costs too many story points from the team's opinion, thus needs a decomposition.

The outcome of this investigation should be:
- [ ] Describing and understanding what are the cases might be, when message ID is unexpected. This should give enough technical details to deduct what and where exactly we have to change in framework . This ticket can be considered as a source of real observations https://github.com/vaadin/flow/issues/12640, but not only limited to those. 
- [ ] A list of enhancements with <= 8 story points that could address the cases, where the network is a culprit, i.e. when the messages are lost. The cases, caused by misconfiguration of session replication, should be described, but out of scope of further enhancements in framework. Though they should be properly documented.
- [ ] A testing procedure needs to be ready, at least the idea how we can test these enhancements and guarantee that they cover real cases.

**Comment by caalador** [comment]
Cases seen in reported logs:
- `Ignore old message from client`
  - Server receives previous message id with the same content hash as last processed
  - Server should resend last response to sync state on client.
- `Unexpected id from client`
  - If the client id is bigger than the expected there has been a dropped client request. (client updates the id on `preparePayload`)
  - Client should be requested to resend previous message and server should sync the id to current request id.
- Server expected much bigger than the client request?
  - Resynchronization should probably happen as the gap is too big and why is the client requesting with much too old id?

**Comment by caalador** [comment]
As the server always answers the client, the client should probably queue the sends as it does for one single pushPendingMessage when using bidirectional transport, but not for xhr. Also `send(JsonObject payload)` is public and doesn't do any queueing so that `sendInvocationsToServer` would purge them instead of pushing it out immediately where we could possibly get faulty sync id from the client.

**Comment by mshabarov** [comment]
Tickets created based on above in priority:
- [ ] https://github.com/vaadin/flow/issues/20507
- [ ] https://github.com/vaadin/flow/issues/20506
- [ ] https://github.com/vaadin/docs/issues/1776
- [ ] TBD - when resync is needed, Flow to give an explanatory message why this is happening and give a link to online docs about reverse proxy/session replication docs.
- [ ] https://github.com/vaadin/flow/issues/4746

**Comment by mshabarov** [comment]
For testing I'd propose to not use any heavy third-party tools, but instead to intercept the request/response right away in a Vaadin applications.

1. Using a custom `AtmosphereInterceptor` to test cases with Vaadin Push / WebSockets:

`inspect(AtmosphereResource r)` — Called when a client sends a message to the server.
`postInspect(AtmosphereResource r)` — Called when a server sends a message to the client.

An example:

```java
import org.atmosphere.config.managed.AtmosphereInterceptorAdapter;
import org.atmosphere.cpr.AtmosphereResource;
import org.atmosphere.cpr.AtmosphereResponse;
import org.atmosphere.cpr.AtmosphereRequest;

public class CustomWebSocketInterceptor extends AtmosphereInterceptorAdapter {

    @Override
    public Action inspect(AtmosphereResource resource) {
        AtmosphereRequest request = resource.getRequest();

        // Intercept client-to-server messages
        String clientMessage = request.getReader().lines()
                .reduce("", (accumulator, actual) -> accumulator + actual);

        // Simulate dropping a message from the client
        if (clientMessage.contains("something")) {
            return Action.CANCELLED; // Stops further processing
        }
        return Action.CONTINUE;
    }

    @Override
    public void postInspect(AtmosphereResource resource) {
        AtmosphereResponse response = resource.getResponse();

        // Intercept server-to-client messages
        String serverMessage = response.toString();

        // Simulate altering a response message
        if (serverMessage.contains("something")) {
            String modifiedMessage = serverMessage.replace("something", "MODIFIED");
            response.write(modifiedMessage);
        }
    }

    @Override
    public void destroy() {
        // Cleanup if needed
    }
}
```

Then declare the interceptor in `atmosphere.xml`:

```xml
<atmosphere-handlers>
    <mapping path="/*" support-session="true" broadcaster-cache="org.atmosphere.cache.UUIDBroadcasterCache">
        <interceptor>com.example.CustomWebSocketInterceptor</interceptor>
    </mapping>
</atmosphere-handlers>
```
What this can give us:
- Drop client-to-server or server-to-client messages to simulate network issues.
- Modify the content of messages dynamically, e.g. to change the clientId/syncId.
- Delay message processing to simulate slow networks.

2. For communication over http, we can use Http Filter API for intercepting requests from client and JS customisations in browser to intercept the response:

```java
@WebFilter("/*")
public class NetworkSimulatingFilter implements Filter {
    @Override
    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain)
            throws IOException, ServletException {
        // Simulate network delay
        Thread.sleep(500);

        // Randomly drop requests
        if (Math.random() < 0.1) { // 10% chance to drop
            ((HttpServletResponse) response).sendError(HttpServletResponse.SC_GATEWAY_TIMEOUT, "Simulated timeout");
            return;
        }

        // Proceed normally
        chain.doFilter(request, response);
    }
}
```

To intercept the Vaadin requests in browser, the following hack for `fetch` may work:

```javascript
(function() {
  const originalFetch = window.fetch;

  window.fetch = function(...args) {
    console.log('Intercepted request:', args);

    // Simulate a network failure
    if (args[0].includes('some-endpoint')) {
      return Promise.reject(new Error('Simulated network error'));
    }

    return originalFetch.apply(this, args).then(response => {
      console.log('Intercepted response:', response);

      // Optionally, clone and modify the response
      if (response.url.includes('another-endpoint')) {
        return response.text().then(body => {
          const modifiedBody = body.replace('original-content', 'modified-content');
          return new Response(modifiedBody, response);
        });
      }

      return response;
    });
  };
})();

```
